### ISSUE TITLE
`no-this-outside-of-class`: allow using `this` as parameter

### ISSUE THREAD

**Issue by krislintigo** [sindresorhus/eslint-plugin-unicorn#3593]
### Description

Currently rule does not allow any this usage outside of classes, though it's required sometimes for Typescript methods that rely on `this` as a first function parameter:

```ts
@Prop({
    type: () => Date,
    default: null,
    validate: {
      validator(this: TrackedModel, value: Date | null) {
        const requiresSpecificDate = this.notificationFrequency === 'once'
        const hasSpecificDate = value !== null

        return requiresSpecificDate === hasSpecificDate
      },
      message: 'notificationDate is required only when notificationFrequency is set to once',
    },
  })
  notificationDate!: Date | null
```

### Examples

```ts
// ❌ requires `eslint-disable`
validator(this: TrackedModel, value: Date | null) {
    // eslint-disable-next-line unicorn/no-this-outside-of-class
    const requiresSpecificDate = this.notificationFrequency === 'once'
    const hasSpecificDate = value !== null

    return requiresSpecificDate === hasSpecificDate
},
```

```ts
// ✅ should pass
validator(this: TrackedModel, value: Date | null) {
    const requiresSpecificDate = this.notificationFrequency === 'once'
    const hasSpecificDate = value !== null

    return requiresSpecificDate === hasSpecificDate
},
```
